File: scripts/read_HY_2b.m
%% read and plot HY-2B data
clear;clc;

%% load HY-2B data
data_dir = '../data/';  % base folder save HY_2B data
files = dir(fullfile(data_dir, '*.h5')); % filenames of the HY-2b data
step = 2;  %original data is too large, here we set "2" to load data every 2 points 

% Initialize variables to save the scatter data
lon_all = [];
lat_all = [];
speed_all = [];
direction_all = [];

% Loop through each file to read data and process
for i = 1:length(files)
    file = fullfile(data_dir, files(i).name); % path of the HY-2B data
    
    % Read data from HDF5 file
    lon = double(h5read(file, '/wvc_lon'));
    lat = double(h5read(file, '/wvc_lat'));
    speed = double(h5read(file, '/wind_speed_selection'));
    direction = double(h5read(file, '/wind_dir_selection'));
    
    % Subsample and reshape data
    lon = lon(1:step:end, 1:step:end);
    lat = lat(1:step:end, 1:step:end);
    speed = speed(1:step:end, 1:step:end) / 100;  % Scale wind speed to m/s
    direction = direction(1:step:end, 1:step:end) / 10 * pi / 180; % Convert to radians
    
    % Reshape to 1D
    lon_1d = reshape(lon, [], 1);
    lat_1d = reshape(lat, [], 1);
    speed_1d = reshape(speed, [], 1);
    direction_1d = reshape(direction, [], 1);
    
    % Remove invalid data where speed < 0
    valid_idx = speed_1d >= 0;
    lon_1d = lon_1d(valid_idx);
    lat_1d = lat_1d(valid_idx);
    speed_1d = speed_1d(valid_idx);
    direction_1d = direction_1d(valid_idx);
    
    % Concatenate data
    lon_all = [lon_all; lon_1d];
    lat_all = [lat_all; lat_1d];
    speed_all = [speed_all; speed_1d];
    direction_all = [direction_all; direction_1d];
end

%% Figure to show the HY-2B data
figure('unit','centimeters','position',[1,1,18,10],'color','w'); % figure size
left = 1.5/18; 
bottom = 1/10;
width = 15/18;
height = 8/10;
axes('position',[left bottom width height]); % axe location
% using m_map to plot the fgure
m_proj('Equidistant cylindrical','long',[0 360],'lat',[-90 90]);
m_scatter(lon_all,lat_all,2,speed_all, 'filled');hold on; 
colormap('jet'); % set 'jet' type of the colormap
caxis([3 15]); % min and max limitaton of the figure
m_coast('patch',[.86 .86 .86]); hold on; % land patch
% vector of the wind direction
x_comp = sin(direction_all);
y_comp = cos(direction_all);
m_quiver(lon_all(1:40:end), lat_all(1:40:end), x_comp(1:40:end), y_comp(1:40:end), 3, ...
        'k', 'MaxHeadSize', 5, 'AutoScale', 'off');
% tick and ticklabels of the x and y coordinate
m_grid('linestyle','none','tickdir','out','xtick',0:60:360,'ytick',-90:30:90,'fontsize',12,...
    'fontname','Times New Roman','linewidth',1.5);
hc = colorbar;
set(hc,'tickdir','out','position',[0.93 0.15 0.012 0.7],...
   'ytick',3:3:15,'fontsize',12,'fontname','Times New Roman');
% export_fig(sprintf('%s%d.png','../figures/HY_2B_matlab'),'-r300','-zbuffer');

File: scripts/read_HY_2B.py
# %% import libraries
import os
import numpy as np
import xarray as xr
import h5py
import datetime
import matplotlib.pyplot as plt
import cartopy.crs as ccrs

"""
Python codes for data analysis of the HY-2B satellite data. 
"""
# %% read filename
file_names = [f for f in os.listdir("../data/") if f.endswith(".h5")]

# %% read attributes about the h5 file
ds = h5py.File(f"../data/{file_names[0]}", "r")
# check the attributes of the h5 file
for k in ds.attrs.keys():
    print("{}=>{}".format(k, ds.attrs[k]))
# check the keys of the h5 file
for key in ds.keys():
    print("\n")
    print(ds[key], key, ds[key].name)
# check the attributes of the keys
for k in ds[key].attrs.keys():
    print(k, ds[key].attrs[k])

# %%
# read all data and concatenate them
lon_all = []
lat_all = []
speed_all = []
direction_all = []

step = 2  # subtract the data every 3 steps

for file in file_names:
    ds = h5py.File(f"../data/{file}", "r")
    lon = ds["wvc_lon"][:][::step, ::step]  # lon
    lat = ds["wvc_lat"][:][::step, ::step]  # lat
    speed = ds["wind_speed_selection"][:][::step, ::step]  # wind speed
    direction = ds["wind_dir_selection"][:][::step, ::step]  # wind direction
    lon_1d = lon.reshape(lon.shape[0] * lon.shape[1], order="F")
    lat_1d = lat.reshape(lat.shape[0] * lat.shape[1], order="F")
    speed_1d = (
        speed.reshape(speed.shape[0] * speed.shape[1], order="F") / 100
    )  # rescale the speed to the actual m/s
    direction_1d = (
        direction.reshape(direction.shape[0] * direction.shape[1], order="F") / 10
    )  # rescale the direction to the actual degree
    direction_1d = (
        np.array(direction_1d) / 360 * 2 * np.pi
    )  # convert the degree to radian
    del_ind = np.where(speed_1d < 0)  # delete the none data
    lon_1d = np.delete(lon_1d, del_ind)
    lat_1d = np.delete(lat_1d, del_ind)
    speed_1d = np.delete(speed_1d, del_ind)
    direction_1d = np.delete(direction_1d, del_ind)
    # concatenate the data
    lon_all = np.concatenate((lon_all, lon_1d))
    lat_all = np.concatenate((lat_all, lat_1d))
    speed_all = np.concatenate((speed_all, speed_1d))
    direction_all = np.concatenate((direction_all, direction_1d))

# %% plot the data
fig = plt.figure(figsize=(18, 10), dpi=300)
ax = plt.axes(projection=ccrs.PlateCarree(central_longitude=180))
ax.coastlines()
scatter = ax.scatter(
    lon_all,
    lat_all,
    s=0.5,
    c=speed_all,
    vmin=3,
    vmax=15,
    cmap="coolwarm",
    transform=ccrs.PlateCarree(),
)
# plot the wind direction
x_comp = np.sin(direction_all)
y_comp = np.cos(direction_all)
ax.quiver(
    lon_all[::40],
    lat_all[::40],
    x_comp[::40],
    y_comp[::40],
    scale=80,
    width=0.001,
    headlength=8,
    headwidth=8,
    transform=ccrs.PlateCarree(),
)
cbar = plt.colorbar(
    scatter,
    ax=ax,
    orientation="vertical",
    extend="both",
    pad=0.01,
    shrink=0.6,
    label="Wind Speed (m/s)",
    ticks=np.arange(3, 16, 3),
)
cbar.ax.tick_params(labelsize=15)
ax.set_xticks(np.arange(0, 360, 30), crs=ccrs.PlateCarree())
ax.set_yticks(np.arange(-90, 91, 30), crs=ccrs.PlateCarree())
lon_labels = [str(int(lon)) for lon in np.arange(0, 360, 30)]
ax.set_xticklabels(lon_labels, fontsize=15)
ax.tick_params(axis="both", labelsize=15)

plt.savefig("../figures/HY_2B_python.png", dpi=300, bbox_inches="tight")
plt.show()
